### PR TITLE
[wasm] Add metadata registration for WebAssembly

### PR DESCRIPTION
**Pull request by kateinoigakukun**
This patch adds the metadata registration for the wasm targets, and also adds build support for it.

### CODE REVIEW

**Comment by kateinoigakukun** [comment]
@swift-ci Please smoke test

File: stdlib/public/runtime/SwiftRT-WASM.cpp
Comment (grynspan) [inline stdlib/public/runtime/SwiftRT-WASM.cpp:58]: Totally appropriate to just leave the base address as `0` if it is meaningless. :)

**Comment by grynspan** [comment]
Can anything here be deduped from the ELF implementation? They two implementations are _very_ similar.

File: stdlib/public/runtime/SwiftRT-WASM.cpp
Comment (kateinoigakukun) [reply stdlib/public/runtime/SwiftRT-WASM.cpp:58]: Thank you for clarification!

Comment (grynspan) [reply stdlib/public/runtime/SwiftRT-WASM.cpp:58]: Happy to help! You might want to consider using `nullptr` instead of `0` for clarity that this is a pointer, not an integer.

**Comment by kateinoigakukun** [comment]
Good point! I think the list of section names can be shared by X-macro.

File: stdlib/public/runtime/SwiftRT-WASM.cpp
Comment (kateinoigakukun) [reply stdlib/public/runtime/SwiftRT-WASM.cpp]: I updated to use `nullptr` explicitly

**Comment by kateinoigakukun** [comment]
@grynspan Thank you for feedback and sorry for my slow response (I was sick last week.. 😵)
I updated my patch to share the most of the content among ELF and Wasm versions :)

**Comment by kateinoigakukun** [comment]
@swift-ci Please smoke test

**Comment by grynspan** [comment]
This looks really good to me! Please be sure to get a review from the core Swift team before merging. :)

**Review by grynspan — APPROVED** [review]

**Review by al45tair — APPROVED** [review]
LGTM.

I wasn't 100% sure about the ELF and WASM versions sharing the `.cpp` file, but it looks like the implementation is very similar so I can see why you did that.

**Comment by kateinoigakukun** [comment]
Thank you all for your kind reviews!

**Comment by grynspan** [comment]
> LGTM.
> 
> I wasn't 100% sure about the ELF and WASM versions sharing the `.cpp` file, but it looks like the implementation is very similar so I can see why you did that.

We can actually probably merge the COFF implementation in too, and then this file gets renamed "SwiftRT.cpp" and world peace is achieved.

**Pull request merged**
